File: ios/objectdetection/ObjectDetectionFrameProcessorPlugin.m
//
//  FrameProcessorPlugins.m
//  react-native-mediapipe
//
//  Created by Charles Parker on 3/24/24.
//

#import <Foundation/Foundation.h>

#import <VisionCamera/FrameProcessorPlugin.h>
#import <VisionCamera/FrameProcessorPluginRegistry.h>

#import "ReactNativeMediaPipe-Swift.h"

@interface ObjectDetectionFrameProcessorPlugin (FrameProcessorPluginLoader)
@end

@implementation ObjectDetectionFrameProcessorPlugin (FrameProcessorPluginLoader)
+ (void) load {
  [FrameProcessorPluginRegistry addFrameProcessorPlugin:@"objectDetection"
                                        withInitializer:^FrameProcessorPlugin*(VisionCameraProxyHolder* proxy, NSDictionary* options) {
    return [[ObjectDetectionFrameProcessorPlugin alloc] initWithProxy:proxy withOptions:options];
  }];
}
@end

File: ios/objectdetection/ObjectDetectionFrameProcessorPlugin.swift
//
//  ObjectDetectionFrameProcessorPlugin.swift
//  react-native-mediapipe
//
//  Created by Charles Parker on 3/24/24.
//

import Foundation
import Vision


@objc(ObjectDetectionFrameProcessorPlugin)
public class ObjectDetectionFrameProcessorPlugin: FrameProcessorPlugin {
  public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]! = [:]) {
    super.init(proxy: proxy, options: options)
  }

  public override func callback(_ frame: Frame, withArguments arguments: [AnyHashable: Any]?) -> Any
  {
    let detectorHandle: Double = arguments?["detectorHandle"] as! Double
    guard let detector = ObjectDetectionModule.detectorMap[Int(detectorHandle)] else {
      return false
    }
  
    let buffer = frame.buffer
    detector.detectAsync(
      sampleBuffer: buffer,
      orientation: frame.orientation,
      timeStamps: Int(Date().timeIntervalSince1970 * 1000))
    return true
  }
}
